### PR TITLE
feat: expose window.invalidateShadow()

### PR DESCRIPTION
**Pull request by codebytere**
#### Description of Change

Closes https://github.com/electron/electron/issues/32450.

This PR exposes a new macOS-only method `window.invalidateShadow()`. This method can be used to clear residual visual artifacts when, for example, performing an animation.

`BrowserWindows` that are transparent and frameless can sometimes leave behind visual artifacts on macOS. To remedy this, the shadow can be forcibly cleared on an NSWindow by calling `-[NSWindow invalidateShadow]`, but within the constraints of how animations are performed e.g. in `webContents` right now, the only way to solve this within the codebase without exposing a new method would be to override `- (void)drawRect:(NSRect)rect`. This has _significant_ negative perf implications - even if we only invalidated the shadow for frameless windows simply the act of calling this method for every single redraw would slow down applications tenfold. As such, the simplest way to remedy this issue is to expose this method to end-users and allow them to invoke it according to their needs. 

However, if someone has an alternate approach for this i'd be more than open to it!

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Exposed `window.invalidateShadow()` to clear residual visual artifacts on macOS.

### CODE REVIEW

**Comment by nornagon** [comment]
Hmm, this seems quite obscure and is fairly easily achievable with the `objc` module, e.g.

```
const objc = require('objc')
const nativeWindowHandle = mainWindow.getNativeWindowHandle().readPointer()
const nsview = objc.wrap(nativeWindowHandle)
nsview.window().invalidateShadow()
```

I'm not necessarily against adding this API but I'd like to understand a bit more about whether we can avoid adding it!

File: shell/browser/native_window.h [hunk 1]
```diff
@@ -159,6 +159,7 @@ class NativeWindow : public base::SupportsUserData,
   virtual bool IsTabletMode() const;
   virtual void SetBackgroundColor(SkColor color) = 0;
   virtual SkColor GetBackgroundColor() = 0;
+  virtual void InvalidateShadow() = 0;
```
Comment (zcbenz) [inline shell/browser/native_window.h:162]: The code fails to compile on Linux/Windows do to lack of implementation. Maybe wrap it in `#if defined(OS_MAC)`?

**Review by zcbenz — APPROVED** [review]
I'm good with this API since a large number of BrowserWindow's APIs are just simple wrappers of native APIs.

**Comment by jkleinsc** [comment]
@codebytere can you rebase this?  that should get rid of the linux failure

**Comment by codebytere** [comment]
@jkleinsc done!

**Review by samuelmaddock — COMMENTED** [review]
I've noticed this before when moving a window between screens. Could we call this internally at the end of dragging/moving a window to solve some amount of problems without requiring developers to know about this method?

**Comment by nornagon** [comment]
Let's investigate whether there's a way we can automatically behave correctly instead of requiring the user to give this signal.

**Review by RaisinTen — COMMENTED** [review]
How often are users expected to call this method during an animation? If it is required for every redraw, I think it's something that should be done internally.

**Comment by nornagon** [comment]
I haven't been able to find a way that this could be automatically triggered, so API LGTM.

**Review by jkleinsc — APPROVED** [review]
API LGTM

File: docs/api/browser-window.md [hunk 2]
```diff
@@ -1482,6 +1482,13 @@ screen readers
 Sets a 16 x 16 pixel overlay onto the current taskbar icon, usually used to
 convey some sort of application status or to passively notify the user.
 
+#### `win.invalidateShadow()` _macOS_
+
+Invalidates the window shadow so that it is recomputed based on the current window shape.
+
+`BrowserWindows` that are transparent and frameless can sometimes leave behind visual artifacts on macOS.
```
[the remark this answers is not in the document]
Comment (RaisinTen) [reply docs/api/browser-window.md:1523]: @nornagon both look the same, right? So why would it not be incorrect for a framed window if it is incorrect for a frameless window?

### Without frame

Uses https://gist.github.com/lyswhut/5112851712bcd11e7045921fee1cbbf7:

https://user-images.githubusercontent.com/42526976/160763766-80ee00ff-6531-4558-8899-d957cf5703b0.mov

### With frame

Uses https://gist.github.com/lyswhut/5112851712bcd11e7045921fee1cbbf7 + this patch:
```diff
diff --git a/main.js b/main.js
index d35ca9f..a596ef4 100644
--- a/main.js
+++ b/main.js
@@ -9,7 +9,7 @@ function createWindow () {
   mainWindow = new BrowserWindow({
     height: 500,
     width: 500,
-    frame: false,
+    frame: true,
     transparent: true,
     show: false,
     alwaysOnTop: true,
```

https://user-images.githubusercontent.com/42526976/160764032-9290a4da-e901-4fa3-bac0-70a189d7df06.mov

**Comment by RaisinTen** [comment]
> I haven't been able to find a way that this could be automatically triggered, so API LGTM.

@nornagon why can't we call this internally whenever users are expected to call it? I'm not sure I fully understand when / how frequently users are expected to call this though, could you please explain that?

**Comment by nornagon** [comment]
> > I haven't been able to find a way that this could be automatically triggered, so API LGTM.
> 
> @nornagon why can't we call this internally whenever users are expected to call it? I'm not sure I fully understand when / how frequently users are expected to call this though, could you please explain that?

Users should call this method when the outline of their window has changed. The vast majority of apps never need to call it.

File: docs/api/browser-window.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (nornagon) [reply docs/api/browser-window.md:1523]: Huh weird, somehow the titlebar is also transparent? That doesn't seem very useful 😂 but sure, yeah the transparency is the important part here, not the frame per se.

**Comment by RaisinTen** [comment]
> Users should call this method when the outline of their window has changed. The vast majority of apps never need to call it.

@nornagon then we should be able to run an event handler when the outline of the window changes only for transparent windows that just calls this function. That way, we can abstract away this functionality from our users.

**Comment by nornagon** [comment]
> > Users should call this method when the outline of their window has changed. The vast majority of apps never need to call it.
> 
> @nornagon then we should be able to run an event handler when the outline of the window changes only for transparent windows that just calls this function. That way, we can abstract away this functionality from our users.

Yep! The only catch is, there's no such event. Chromium doesn't tell us when the outline of the window changes.

**Comment by RaisinTen** [comment]
> > > Users should call this method when the outline of their window has changed. The vast majority of apps never need to call it.
> > 
> > 
> > @nornagon then we should be able to run an event handler when the outline of the window changes only for transparent windows that just calls this function. That way, we can abstract away this functionality from our users.
> 
> Yep! The only catch is, there's no such event, as far as I can tell. Chromium doesn't tell us when the outline of the window changes.

@nornagon ah, but then how are users going to figure out when this API should be called if we can't expose a listener for the outline changing event?

**Comment by nornagon** [comment]
@RaisinTen presumably the app changes its outline for some reason; the app can signal the browser when that change happens from the app logic.

**Comment by RaisinTen** [comment]
@nornagon

> Yep! The only catch is, there's no such event, as far as I can tell. Chromium doesn't tell us when the outline of the window changes.

Aren't the [`'move'`](https://www.electronjs.org/docs/latest/api/browser-window#event-move) and [`'resize'`](https://www.electronjs.org/docs/latest/api/browser-window#event-resize) events enough for this? I'm assuming that outline = bounds of a window.

> @RaisinTen presumably the app changes its outline for some reason; the app can signal the browser when that change happens from the app logic.

Are you talking about the use case where the app changes the outline of a window programmatically alone?

Btw, I tested locally that both the [`'move'`](https://www.electronjs.org/docs/latest/api/browser-window#event-move) and [`'resize'`](https://www.electronjs.org/docs/latest/api/browser-window#event-resize) events get triggered when the window outline is changed both programmatically as well as when it is changed manually by the consumers of the app.

**Comment by RaisinTen** [comment]
Actually, no. Just calling `invalidateShadow()` when the window outline changes is not enough. It still draws the shadow after the window has been moved / resized. So I'm still not sure how users are supposed to use it. 🤔 

### Without `invalidateShadow()`

Uses https://gist.github.com/lyswhut/5112851712bcd11e7045921fee1cbbf7:

https://user-images.githubusercontent.com/42526976/161257872-dea566ef-def6-4226-ba79-51d83fe8a71f.mov

### With `invalidateShadow()`

Uses https://gist.github.com/lyswhut/5112851712bcd11e7045921fee1cbbf7 + this patch:

```diff
diff --git a/main.js b/main.js
index ab85d6b..0ff4de3 100644
--- a/main.js
+++ b/main.js
@@ -20,6 +20,9 @@ function createWindow () {
     },
   })

+  mainWindow.on('resize', mainWindow.invalidateShadow)
+  mainWindow.on('move', mainWindow.invalidateShadow)
+
   mainWindow.once('ready-to-show', () => {
     mainWindow.show()
   })
```

https://user-images.githubusercontent.com/42526976/161258228-063c3c0a-56cc-476e-8f70-091a0eb8b7c0.mov

**Review by ckerr — COMMENTED** [review]
API LGTM

**Comment by xiajingren** [comment]
Any progress?

**Comment by nornagon** [comment]
@RaisinTen you've misunderstood when the function needs to be called; it's not just when the window is moved or resized, but whenever the _outline_ of the window changes. in this example, that happens whenever the number is updated, or when the window scrolls.

**Review by nornagon — APPROVED** [review]
API LGTM

**Comment by nornagon** [comment]
I think this is reasonable to land; we've looked for a different way to do it and haven't found one. It makes sense to allow apps to invalidate their window shadows when they know the window shape has changed.

**Review by RaisinTen — APPROVED** [review]

**Comment by lyswhut** [comment]
Can this be merged into 20.xy and 21.xy ?

**Comment by lyswhut** [comment]
@codebytere Can this be merged into 20.xy and 21.xy ?

**Comment by lyswhut** [comment]
@codebytere @RaisinTen Electron 23 no longer supports Windows 7, can this PR be merged into 22.X? I want to try this API :)

**Pull request merged**